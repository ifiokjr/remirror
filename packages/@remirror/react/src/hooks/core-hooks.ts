import isEqual from 'fast-deep-equal/react';
import {
  DependencyList,
  RefObject,
  useCallback,
  useEffect,
  useLayoutEffect,
  useRef,
  useState,
} from 'react';
import useEffectOnce from 'react-use/lib/useEffectOnce';
import useSetState from 'react-use/lib/useSetState';
import useUpdateEffect from 'react-use/lib/useUpdateEffect';
import ResizeObserver from 'resize-observer-polyfill';

import { object } from '@remirror/core';

/**
 * Returning a new object reference guarantees that a before-and-after
 * equivalence check will always be false, resulting in a re-render, even when
 * multiple calls to forceUpdate are batched.
 */
export function useForceUpdate(): () => void {
  const [, setState] = useState(object());

  const forceUpdate = useCallback((): void => {
    setState(object());
  }, []);

  return forceUpdate;
}

/**
 * Preserves the previous version of a provided value.
 *
 * ```tsx
 * const [isOpen, setOpen] = useState<boolean>(false)
 * const previous = usePrevious(isOpen)
 *
 * return <span onClick={() => setOpen(!isOpen)}>{isOpen && previous === isOpen ? 'Stable' : 'Unstable' }</span>
 * ```
 */
export function usePrevious<Value>(value: Value): Value | undefined {
  const ref = useRef<Value>();
  useEffect(() => void (ref.current = value), [value]);
  return ref.current;
}

export interface DOMRectReadOnlyLike {
  readonly x: number;
  readonly y: number;
  readonly width: number;
  readonly height: number;
  readonly top: number;
  readonly right: number;
  readonly bottom: number;
  readonly left: number;
}

const defaultBounds = { x: 0, y: 0, width: 0, height: 0, top: 0, right: 0, bottom: 0, left: 0 };

export type UseMeasureReturn<Ref extends HTMLElement = HTMLElement> = readonly [
  { readonly ref: RefObject<Ref> },
  DOMRectReadOnlyLike,
];

/**
 * Provides the measurements for a react element at the point of layout.
 *
 * @remarks
 *
 * Taken from https://codesandbox.io/embed/lp80n9z7v9
 *
 * ```tsx
 * const [bindRef, { height }] = useMeasure()
 *
 * return <div {...bindRef}>Height: {height}</div>
 * ```
 */
export function useMeasure<Ref extends HTMLElement = HTMLElement>(): UseMeasureReturn<Ref> {
  const ref = useRef<Ref>(null);
  const [bounds, setBounds] = useState<DOMRectReadOnlyLike>(defaultBounds);

  useLayoutEffect(() => {
    const resizeObserver = new ResizeObserver(([entry]) => setBounds(entry.contentRect));

    if (ref.current) {
      resizeObserver.observe(ref.current);
    }

    return () => resizeObserver.disconnect();
  }, []);

  return [{ ref }, bounds] as const;
}

/**
 * The `setState` type for the `useSetState` hook.
 */
export type PartialDispatch<Type extends object> = (
  patch: Partial<Type> | ((prevState: Type) => Partial<Type>),
) => void;

/**
 * A `useEffect` function which issues a warning when the dependencies provided
 * are deeply equal, but only in development.
 *
 * This is used in places where it's important for developers to memoize and
 * wrap methods with `useCallback`.
 */
const useEffectWithWarning: typeof useEffect =
  process.env.NODE_ENV === 'production'
    ? useEffect
    : (effect, deps) => {
        const ref = useRef<DependencyList>();
        const unnecessaryChange = useRef(0);

        useEffectOnce(() => {
          ref.current = deps;
        });

        useUpdateEffect(() => {
          if (!isEqual(deps, ref.current)) {
            unnecessaryChange.current = 0;
            ref.current = deps;
            return;
          }

          unnecessaryChange.current += 1;
        });

        const wrappedEffect = () => {
          if (unnecessaryChange.current >= 1) {
            console.warn(
              `The dependencies passed into your useEffect are deeply equal, but an update has been triggered ${unnecessaryChange.current} time(s). Please consider wrapping the values with \`useMemo\` or \`useCallback\` to memoize your dependencies and prevent unnecessary re-renders.`,
              deps,
            );
          }

          return effect();
        };

        // eslint-disable-next-line react-hooks/exhaustive-deps
        useEffect(wrappedEffect, deps);
      };

export { useEffectWithWarning, useSetState };
