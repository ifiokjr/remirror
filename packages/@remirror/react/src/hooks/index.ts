export * from './core-hooks';
export * from './editor-hooks';
export * from './use-react-editor';
