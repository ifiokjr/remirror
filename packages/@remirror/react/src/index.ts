export type { I18nProviderProps, RemirrorProviderProps, ThemeProviderProps } from './components';
export { I18nProvider, RemirrorProvider, ThemeProvider } from './components';

export type {
  BaseReactCombinedUnion,
  DOMRectReadOnlyLike,
  UseExtensionCallback,
  UsePositionerReturn,
  UseMultiPositionerReturn,
  UseRemirrorType,
} from './hooks';
export {
  useEffectWithWarning,
  useExtension,
  useForceUpdate,
  useI18n,
  useManager,
  useMeasure,
  useMultiPositioner,
  usePositioner,
  usePreset,
  usePrevious,
  useRemirror,
} from './hooks';

export { createReactManager } from './react-helpers';

export type {
  BaseProps,
  CreateReactManagerOptions,
  DefaultReactCombined,
  GetRootPropsConfig,
  I18nContextProps,
  ReactCombinedUnion,
  RefKeyRootProps,
  RefParameter,
  RemirrorContextProps,
} from './react-types';

// Export boundary already applied
export * from './renderers';
