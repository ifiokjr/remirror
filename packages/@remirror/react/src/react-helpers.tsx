import { AnyCombinedUnion, getLazyArray, isRemirrorManager, RemirrorManager } from '@remirror/core';
import { CorePreset } from '@remirror/preset-core';
import { ReactPreset } from '@remirror/preset-react';

import type { CreateReactManagerOptions, ReactCombinedUnion } from './react-types';

/**
 * Create a react manager with all the default react presets and extensions.
 */
export function createReactManager<Combined extends AnyCombinedUnion>(
  combined: Combined[] | (() => Combined[]) | RemirrorManager<ReactCombinedUnion<Combined>>,
  options: CreateReactManagerOptions = {},
): RemirrorManager<ReactCombinedUnion<Combined>> {
  const { core, react, ...settings } = options;

  if (isRemirrorManager<ReactCombinedUnion<Combined>>(combined)) {
    return combined;
  }

  return RemirrorManager.create(
    () => [...getLazyArray(combined), new ReactPreset(react), new CorePreset(core)],
    settings,
  );
}
