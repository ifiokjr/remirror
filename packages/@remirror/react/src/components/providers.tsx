import type { ReactChild, ReactFragment } from 'react';
import React, {
  CSSProperties,
  ElementType,
  ProviderProps,
  ReactElement,
  ReactNode,
  ReactPortal,
  useEffect,
} from 'react';

import type { AnyCombinedUnion, MakeOptional } from '@remirror/core';
import { i18n as defaultI18n } from '@remirror/i18n';
import { RemirrorType } from '@remirror/react-utils';
import type { RemirrorThemeType } from '@remirror/theme';
import { createThemeVariables, themeStyles } from '@remirror/theme';

import { useManager } from '../hooks';
import { I18nContext, RemirrorContext } from '../react-contexts';
import type {
  BaseProps,
  CreateReactManagerOptions,
  GetRootPropsConfig,
  I18nContextProps,
  RemirrorContextProps,
} from '../react-types';
import { ReactEditor } from './react-editor';

interface RemirrorContextProviderProps<Combined extends AnyCombinedUnion>
  extends ProviderProps<RemirrorContextProps<Combined>>,
    Pick<RemirrorProviderProps<Combined>, 'childAsRoot'> {}

export interface RemirrorProviderProps<Combined extends AnyCombinedUnion>
  extends MakeOptional<BaseProps<Combined>, 'manager'> {
  /**
   * The `RemirrorProvider` only supports **ONE** child element. You can place
   * the child element in a fragment if needed.
   */
  children: ReactChild | ReactFragment | ReactPortal;

  /**
   * The presets and extensions that you would like to use to automatically
   * create the manager.
   *
   * These extensions and presets can't be updated and are not dynamic. You
   * should not change they during the component lifecycle as they are created
   * once at the very start and never recreated.
   */
  combined?: Combined[];

  /**
   * The settings to provide to the `RemirrorManager`, `ReactPreset` and
   * `CorePreset`. This is only applied when no `manager` is provided.
   */
  settings?: CreateReactManagerOptions;

  /**
   * Sets the first child element as a the root (where the prosemirror editor
   * instance will be rendered).
   *
   * @remarks
   *
   * **Example with directly nested components**
   *
   * When using a remirror provider calling `getRootProps` is mandatory. By
   * setting `childAsRoot` to an object Remirror will inject these props into
   * the first child element.
   *
   * ```tsx
   * import { RemirrorProvider, useManager } from '@remirror/react';
   *
   * const Editor = () => {
   *   const manager = useManager([...myExtensions]);
   *
   *   return (
   *     <RemirrorProvider childAsRoot={{ refKey: 'ref' }}>
   *       <div />
   *     </RemirrorProvider>,
   *   );
   * };
   * ```
   *
   * If this is set to an empty object then the outer element must be able to
   * receive a default ref prop which will mount the editor to it. If left
   * undefined then the children components are responsible for calling
   * `getRootProps`.
   *
   * @default undefined
   */
  childAsRoot?: GetRootPropsConfig<string> | boolean;
}

/**
 * This purely exists so that we know when the remirror editor has been called
 * with a provider as opposed to directly as a render prop by the user.
 *
 * It's important because, when called directly by the user, `getRootProps` is
 * automatically called when the render prop is called. However, when called via
 * a Provider, the render prop renders the context component and it's not until
 * the element is actually rendered that the getRootProp in any nested
 * components is called.
 */
const RemirrorContextProvider = <Combined extends AnyCombinedUnion>({
  childAsRoot: _,
  ...props
}: RemirrorContextProviderProps<Combined>) => {
  return <RemirrorContext.Provider {...props} />;
};

RemirrorContextProvider.$$remirrorType = RemirrorType.ContextProvider;
RemirrorContextProvider.defaultProps = {
  childAsRoot: false,
};

/**
 * The RemirrorProvider which injects context into it's child component.
 *
 * @remarks
 * This only supports one child. At the moment if that that child is a built in
 * html string element then it is also where the prosemirror editor will be
 * injected (root element).
 *
 * These can either be consumed using React Hooks
 * - `useRemirror`
 * - `usePositioner`
 */
export const RemirrorProvider = <Combined extends AnyCombinedUnion>(
  props: RemirrorProviderProps<Combined>,
): ReactElement<RemirrorProviderProps<Combined>> => {
  const { children, childAsRoot, manager, combined, settings, ...rest } = props;

  return (
    <ReactEditor {...rest} manager={useManager(manager ?? combined ?? [], settings)}>
      {(value) => {
        return (
          <RemirrorContextProvider value={value} childAsRoot={childAsRoot}>
            {children}
          </RemirrorContextProvider>
        );
      }}
    </ReactEditor>
  );
};

RemirrorProvider.$$remirrorType = RemirrorType.Provider;

export interface I18nProviderProps extends Partial<I18nContextProps> {
  children: ReactNode;
}

/**
 * A provider component for the remirror i18n helper library.
 *
 * This uses `@lingui/core` in the background. So please star and support the
 * project when you have a moment.
 */
export const I18nProvider = (props: I18nProviderProps): ReactElement<I18nProviderProps> => {
  const { i18n = defaultI18n, locale = 'en', supportedLocales, children } = props;

  useEffect(() => {
    i18n.activate(locale, supportedLocales ?? [locale]);
  }, [i18n, locale, supportedLocales]);

  return (
    <I18nContext.Provider value={{ i18n, locale, supportedLocales }}>
      {children}
    </I18nContext.Provider>
  );
};

I18nProvider.$$remirrorType = RemirrorType.I18nProvider;

export interface ThemeProviderProps {
  /**
   * The theme to customise the look and feel of your remirror editor.
   */
  theme: RemirrorThemeType;

  /**
   * The custom component to use for rendering this editor.
   *
   * @default 'div'
   */
  as?: ElementType<{ style?: CSSProperties; className?: string }>;

  children: ReactNode;
}

/**
 * This the `ThemeProvider`. Wrap your editor with it to customise the theming
 * of content within your editor.
 *
 * Please be aware that this wraps your component in an extra dom layer.
 */
export const ThemeProvider = (props: ThemeProviderProps): ReactElement<ThemeProviderProps> => {
  const { theme, children, as: Component = 'div' } = props;

  return (
    <Component style={createThemeVariables(theme).styles} className={themeStyles}>
      {children}
    </Component>
  );
};

ThemeProvider.$$remirrorType = RemirrorType.ThemeProvider;
