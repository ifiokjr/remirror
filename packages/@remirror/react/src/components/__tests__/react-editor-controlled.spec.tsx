import { RemirrorTestChain } from 'jest-remirror';
import React, { FC, useState } from 'react';

import {
  AnyCombinedUnion,
  AnyExtension,
  EditorState,
  fromHtml,
  PlainExtension,
  RemirrorEventListener,
  SchemaFromCombined,
  StateUpdateLifecycleParameter,
} from '@remirror/core';
import { BoldExtension, ItalicExtension } from '@remirror/testing';
import { act, fireEvent, render, strictRender } from '@remirror/testing/react';

import { useManager, useRemirror } from '../../hooks';
import { createReactManager } from '../../react-helpers';
import type { RemirrorContextProps } from '../../react-types';
import { RemirrorProvider } from '../providers';
import { ReactEditor } from '../react-editor';

const label = 'Remirror editor';

function create<Combined extends AnyCombinedUnion>(combined?: Combined[]) {
  const manager = createReactManager(combined ?? []);
  const chain = RemirrorTestChain.create(manager);

  return {
    manager,
    chain,
    doc: chain.nodes.doc,
    p: chain.nodes.paragraph,
    props: {
      label,
      stringHandler: fromHtml,
    },
  };
}

let errorSpy = jest.spyOn(console, 'error');

beforeEach(() => {
  errorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
});

afterEach(() => {
  errorSpy.mockRestore();
});

describe('Remirror Controlled Component', () => {
  it('should set the initial value', () => {
    const { manager, props } = create();

    const value = manager.createState({
      content: '<p>This is the initial value</p>',
      stringHandler: fromHtml,
    });
    const onChange = jest.fn();

    const { getByRole } = strictRender(
      <ReactEditor {...props} value={value} manager={manager} onChange={onChange}>
        {() => <div />}
      </ReactEditor>,
    );

    expect(getByRole('textbox')).toMatchSnapshot();

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange.mock.calls[0][0].firstRender).toBeTrue();
  });

  it('overrides initial content', () => {
    const { manager, props, chain } = create();

    const value = manager.createState({
      content: '<p>Not terrible</p>',
      stringHandler: fromHtml,
    });
    const onChange = jest.fn();

    strictRender(
      <ReactEditor
        {...props}
        manager={manager}
        initialContent='<p>Terrible</p>'
        value={value}
        onChange={onChange}
      >
        {() => <div />}
      </ReactEditor>,
    );

    expect(chain.dom).toMatchSnapshot();
  });

  it('responds to updates to the editor state', () => {
    const { manager, props, chain } = create();

    const Component = () => {
      const [value, setValue] = useState<EditorState>(
        manager.createState({
          content: '<p>some content</p>',
          stringHandler: fromHtml,
        }),
      );

      return (
        <ReactEditor
          {...props}
          value={value}
          manager={manager}
          onChange={(parameter) => {
            setValue(parameter.state);
          }}
        >
          {() => <div />}
        </ReactEditor>
      );
    };

    strictRender(<Component />);

    act(() => {
      chain.dispatchCommand(({ dispatch, tr }) => dispatch(tr.insertText('add more value ')));
    });

    expect(chain.dom).toMatchSnapshot();
  });

  it('can override updates to the editor state', () => {
    const { manager, props, chain } = create();

    const Component = () => {
      const [value, setValue] = useState<EditorState>(
        manager.createState({
          content: '<p>some content</p>',
          stringHandler: fromHtml,
        }),
      );

      return (
        <ReactEditor
          {...props}
          value={value}
          manager={manager}
          onChange={(parameter) => {
            const { createStateFromContent, getText } = parameter;

            setValue(createStateFromContent(`<p>Hello</p><p>${getText()}</p>`));
          }}
        >
          {() => <div />}
        </ReactEditor>
      );
    };

    strictRender(<Component />);

    act(() => {
      chain.dispatchCommand(({ dispatch, tr }) => dispatch(tr.insertText('add more value ')));
    });

    expect(chain.dom).toMatchSnapshot();
  });

  it('throws when using  `setContent` updates', () => {
    const { manager, props } = create();

    const value = manager.createState({
      content: '<p>some content</p>',
      stringHandler: fromHtml,
    });

    let ctx: RemirrorContextProps<typeof manager['~EP']>;

    const Component: FC<{
      editorState: EditorState<SchemaFromCombined<typeof manager['~EP']>>;
    }> = ({ editorState }) => {
      return (
        <ReactEditor {...props} value={editorState} manager={manager} onChange={jest.fn()}>
          {(context) => {
            ctx = context as any;
            return <div />;
          }}
        </ReactEditor>
      );
    };

    strictRender(<Component editorState={value} />);

    expect(() => ctx.setContent('<p>Error</p>')).toThrowErrorMatchingSnapshot();
    expect(() => ctx.clearContent()).toThrowErrorMatchingSnapshot();
  });

  it('throws when switching from controlled to non-controlled', () => {
    const { manager, props } = create();

    const value = manager.createState({
      content: '<p>some content</p>',
      stringHandler: fromHtml,
    });

    const set = jest.fn();

    const Component = () => {
      const [state, setState] = useState(value);
      set.mockImplementation(setState);

      return (
        <ReactEditor {...props} value={state} manager={manager} onChange={jest.fn()}>
          {() => <div />}
        </ReactEditor>
      );
    };

    strictRender(<Component />);

    expect(() =>
      act(() => {
        set();
      }),
    ).toThrowErrorMatchingSnapshot();
    expect(errorSpy).toHaveBeenCalled();
  });

  it('throws when switching from non-controlled to controlled', () => {
    const { manager, props } = create();

    const value = manager.createState({
      content: '<p>some content</p>',
      stringHandler: fromHtml,
    });

    const set = jest.fn();

    const Component = () => {
      const [state, setState] = useState();
      set.mockImplementation(setState);

      return (
        <ReactEditor {...props} value={state} manager={manager} onChange={jest.fn()}>
          {() => <div />}
        </ReactEditor>
      );
    };

    strictRender(<Component />);

    expect(() =>
      act(() => {
        set(value);
      }),
    ).toThrowErrorMatchingSnapshot();
    expect(errorSpy).toHaveBeenCalled();
  });

  it('notifies extensions of state updates via `manager.onStateUpdate`', () => {
    const mock = jest.fn();

    class UpdateExtension extends PlainExtension {
      get name() {
        return 'update' as const;
      }

      onStateUpdate: (update: StateUpdateLifecycleParameter) => void = mock;
    }

    const { manager, props, chain, doc, p } = create([new UpdateExtension()]);

    const Component = () => {
      const [value, setValue] = useState<EditorState>(
        manager.createState({
          content: doc(p('some content')),
          stringHandler: fromHtml,
        }),
      );

      return (
        <ReactEditor
          {...props}
          value={value}
          manager={manager}
          onChange={(parameter) => {
            const { state } = parameter;

            setValue(state);
          }}
        >
          {() => <div />}
        </ReactEditor>
      );
    };

    strictRender(<Component />);

    act(() => {
      chain.commands.insertText('First text update');
    });

    expect(mock).toHaveBeenCalledTimes(4);

    const { state, previousState } = mock.mock.calls[3][0];

    expect(state).toBe(chain.state);
    expect(state).not.toBe(previousState);
  });
});

test('can run multiple commands', () => {
  const { manager, props, chain, doc, p } = create([new BoldExtension(), new ItalicExtension()]);
  const { bold, italic } = chain.marks;

  const InnerComponent: FC = () => {
    const { getRootProps, commands } = useRemirror<BoldExtension | ItalicExtension>();

    return (
      <div>
        <div data-testid='target' {...getRootProps()} />
        <button
          onClick={() => {
            commands.toggleBold();
            commands.toggleItalic();
          }}
        />
      </div>
    );
  };

  const Component = () => {
    const [value, setValue] = useState<EditorState>(
      manager.createState({
        content: '',
        stringHandler: fromHtml,
      }),
    );

    return (
      <RemirrorProvider
        {...props}
        value={value}
        manager={manager}
        onChange={(parameter) => {
          const { state } = parameter;
          setValue(state);
        }}
      >
        <InnerComponent />
      </RemirrorProvider>
    );
  };

  const { getByRole } = strictRender(<Component />);

  act(() => {
    chain.commands.insertText('This');
  });

  act(() => {
    chain.selectText('all');
  });

  act(() => {
    fireEvent.click(getByRole('button'));
  });

  expect(chain.state.doc).toEqualRemirrorDocument(doc(p(bold(italic('This')))));
});

test('NOTE: this test is to show that synchronous state updates only show the most recent state update', () => {
  const { manager, props, chain, doc, p } = create([]);

  const InnerComponent: FC = () => {
    const { getRootProps, view } = useRemirror();

    return (
      <div>
        <div data-testid='target' {...getRootProps()} />
        <button
          onClick={() => {
            // TWO UPDATES
            view.dispatch(view.state.tr.insertText('a'));
            view.dispatch(view.state.tr.insertText('b'));
          }}
        />
      </div>
    );
  };

  const Component = () => {
    const [value, setValue] = useState<EditorState>(
      manager.createState({
        content: '',
        stringHandler: fromHtml,
      }),
    );

    return (
      <RemirrorProvider
        {...props}
        value={value}
        manager={manager}
        onChange={(parameter) => {
          const { state } = parameter;
          setValue(state);
        }}
      >
        <InnerComponent />
      </RemirrorProvider>
    );
  };

  const { getByRole } = strictRender(<Component />);

  act(() => {
    fireEvent.click(getByRole('button'));
  });

  expect(chain.state.doc).toEqualRemirrorDocument(doc(p('b')));
});

test('support for rendering a nested controlled editor in strict mode', () => {
  const chain = RemirrorTestChain.create(createReactManager(() => [new BoldExtension()]));

  const Component = () => {
    const manager = useManager(chain.manager);

    const [value, setValue] = useState(
      manager.createState({
        content: '<p>test</p>',
        selection: 'all',
        stringHandler: fromHtml,
      }),
    );

    const onChange: RemirrorEventListener<AnyExtension> = ({ state }) => {
      setValue(state);
    };

    return (
      <RemirrorProvider manager={manager} onChange={onChange} value={value}>
        <div id='1'>
          <TextEditor />
        </div>
      </RemirrorProvider>
    );
  };

  const TextEditor = () => {
    const { getRootProps, active, commands } = useRemirror<BoldExtension>({ autoUpdate: true });

    return (
      <>
        <div {...getRootProps()} />
        <button
          onClick={() => commands.toggleBold()}
          style={{ fontWeight: active.bold() ? 'bold' : undefined }}
        />
      </>
    );
  };

  const { getByRole } = strictRender(<Component />);
  const button = getByRole('button');

  act(() => {
    fireEvent.click(button);
  });

  expect(button).toHaveStyle('font-weight: bold');

  act(() => {
    fireEvent.click(getByRole('button'));
  });

  expect(button).not.toHaveStyle('font-weight: bold');
});

describe('onChange', () => {
  const chain = RemirrorTestChain.create(createReactManager(() => [new BoldExtension()]));
  const mock = jest.fn();

  const Component = () => {
    const manager = useManager(chain.manager);

    const [value, setValue] = useState(
      manager.createState({
        content: '<p>A</p>',
        selection: 'end',
        stringHandler: fromHtml,
      }),
    );

    const onChange: RemirrorEventListener<AnyExtension> = ({ state }) => {
      setValue(state);
      mock(value.doc.textContent);
    };

    return (
      <RemirrorProvider manager={manager} onChange={onChange} value={value}>
        <TextEditor />
      </RemirrorProvider>
    );
  };

  const TextEditor = () => {
    const { getRootProps } = useRemirror<BoldExtension>();

    return (
      <>
        <div {...getRootProps()} />
      </>
    );
  };

  it('updates values', () => {
    render(<Component />);

    for (const char of 'mazing!') {
      act(() => {
        chain.insertText(char);
      });
    }

    expect(mock).toHaveBeenCalledTimes(8);
    expect(mock).toHaveBeenLastCalledWith('Amazing');
  });

  it('updates values in `StrictMode`', () => {
    strictRender(<Component />);

    for (const char of 'mazing!') {
      act(() => {
        chain.insertText(char);
      });
    }

    expect(mock).toHaveBeenCalledTimes(8);
    expect(mock).toHaveBeenLastCalledWith('Amazing');
  });
});
