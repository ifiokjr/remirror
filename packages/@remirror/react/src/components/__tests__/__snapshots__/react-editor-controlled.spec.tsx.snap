// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Remirror Controlled Component can override updates to the editor state 1`] = `
<div
  aria-label="Remirror editor"
  aria-multiline="true"
  aria-placeholder=""
  class="ProseMirror remirror-editor"
  contenteditable="true"
  role="textbox"
>
  <p>
    Hello
  </p>
  <p>
    add more value Hello some content
  </p>
</div>
`;

exports[`Remirror Controlled Component overrides initial content 1`] = `
<div
  aria-label="Remirror editor"
  aria-multiline="true"
  aria-placeholder=""
  class="ProseMirror remirror-editor"
  contenteditable="true"
  role="textbox"
>
  <p>
    Not terrible
  </p>
</div>
`;

exports[`Remirror Controlled Component responds to updates to the editor state 1`] = `
<div
  aria-label="Remirror editor"
  aria-multiline="true"
  aria-placeholder=""
  class="ProseMirror remirror-editor"
  contenteditable="true"
  role="textbox"
>
  <p>
    add more value some content
  </p>
</div>
`;

exports[`Remirror Controlled Component should set the initial value 1`] = `
<div
  aria-label="Remirror editor"
  aria-multiline="true"
  aria-placeholder=""
  class="ProseMirror remirror-editor"
  contenteditable="true"
  role="textbox"
>
  <p>
    This is the initial value
  </p>
</div>
`;

exports[`Remirror Controlled Component throws when switching from controlled to non-controlled 1`] = `
"There is a problem with your controlled editor setup.

You have attempted to switch from a controlled to an uncontrolled editor. Once you set up an editor as a controlled editor it must always provide a \`value\` prop.

For more information visit https://remirror.io/docs/errors#rmr0203"
`;

exports[`Remirror Controlled Component throws when switching from non-controlled to controlled 1`] = `
"There is a problem with your controlled editor setup.

You have provided a \`value\` prop to an uncontrolled editor. In order to set up your editor as controlled you must provide the \`value\` prop from the very first render.

For more information visit https://remirror.io/docs/errors#rmr0203"
`;

exports[`Remirror Controlled Component throws when using  \`setContent\` updates 1`] = `
"There is a problem with your controlled editor setup.

Controlled editors do not support \`clearContent\` or \`setContent\` where \`triggerChange\` is \`true\`. Update the \`value\` prop instead.

For more information visit https://remirror.io/docs/errors#rmr0203"
`;

exports[`Remirror Controlled Component throws when using  \`setContent\` updates 2`] = `
"There is a problem with your controlled editor setup.

Controlled editors do not support \`clearContent\` or \`setContent\` where \`triggerChange\` is \`true\`. Update the \`value\` prop instead.

For more information visit https://remirror.io/docs/errors#rmr0203"
`;
